lesson-003-web-server-graceful-shutdown/main.go

```diff
@@ -0,0 +1,71 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+)
+
+// listenAddr is the address where our web server will be listening for
+// requests.
+const listenAddr = "0.0.0.0:8080"
+
+// homeHandler takes a response writer to build a response for the given
+// request.
+// This http handler will greet you with the hostname of the machine where this
+// app is running.
+func homeHandler(w http.ResponseWriter, r *http.Request) {
+	// We begin by looking up the hostname.
+	host, err := os.Hostname()
+	if err != nil {
+		// If we see an error then we return an http code 500 and we tell the
+		// client what the error was.
+		w.WriteHeader(http.StatusInternalServerError)
+		errMsg := fmt.Sprintf("we saw an error: %v\n", err)
+		fmt.Fprintf(w, errMsg)
+		return
+	}
+
+	// Build a string with the hostname.
+	greeting := fmt.Sprintf("Greeting from %s!\n", host)
+	fmt.Fprintf(w, greeting)
+	time.Sleep(2 * time.Second)
+}
+
+func main() {
+	sigint := make(chan os.Signal, 1)
+
+	// Create a request multiplexer. This will match an incoming request to a
+	// route.
+	mux := http.NewServeMux()
+	server := http.Server{Addr: listenAddr, Handler: mux}
+
+	// Register homeHandler with the router "/". This means that a request to
+	// http://0.0.0.0:8080/ will be handled by the 'homeHandler' function.
+	mux.HandleFunc("/", homeHandler)
+
+	// Interrupt signal (sent from terminal e.g. Ctrl + C)
+	signal.Notify(sigint, os.Interrupt)
+	// SIGTERM signal (sent from kubernetes)
+	signal.Notify(sigint, syscall.SIGTERM)
+
+	go func() {
+		if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			log.Fatal(err)
+		}
+	}()
+
+	<-sigint
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	server.Shutdown(ctx)
```

Comment:
This would only be executed after a termination signal is received.
It may be useful to add another dummy channel ("done") that is closed after the termination signal has been received (to allow the app to finish doing whatever it is doing - give it some additional time to finish its work) or if more signals are received (e.g., second Ctrl+C to indicate that I want to exit right now).

Reply:
Umm that's a really good point, I was thinking to leave this `(to allow the app to finish doing whatever it is doing - give it some additional time to finish its work)` as an extension suggested to users.
But happy to include it in the code proposed if agreed 😄

Reply:
Im thinking we should add it.
letting work finish is a crucial part of a graceful shutdown.
Customizing "how work finishes" then its something that ppl can then experiment with